pkg/audit/manager.go

```diff
@@ -363,34 +387,37 @@ func (am *Manager) getUpdateListsFromAuditResponses(res []*constraintTypes.Resul
 		message := r.Msg
 		resource, ok := r.Resource.(*unstructured.Unstructured)
 		if !ok {
-			return nil, nil, nil, errors.Errorf("could not cast resource as reviewResource: %v", r.Resource)
+			return errors.Errorf("could not cast resource as reviewResource: %v", r.Resource)
 		}
 		rname := resource.GetName()
 		rkind := resource.GetKind()
 		rnamespace := resource.GetNamespace()
-		result := auditResult{
-			cgvk:              gvk,
-			capiversion:       apiVersion,
-			cname:             name,
-			cnamespace:        namespace,
-			rkind:             rkind,
-			rname:             rname,
-			rnamespace:        rnamespace,
-			message:           message,
-			enforcementAction: enforcementAction,
-			constraint:        r.Constraint,
+		// append audit results only if it is below violations limit
+		if uint(len(updateLists[selfLink])) < *constraintViolationsLimit {
+			result := auditResult{
+				cgvk:              gvk,
+				capiversion:       apiVersion,
+				cname:             name,
+				cnamespace:        namespace,
+				rkind:             rkind,
+				rname:             rname,
+				rnamespace:        rnamespace,
+				message:           message,
+				enforcementAction: enforcementAction,
+				constraint:        r.Constraint,
+			}
+			updateLists[selfLink] = append(updateLists[selfLink], result)
 		}
-		updateLists[selfLink] = append(updateLists[selfLink], result)
 		ea := util.EnforcementAction(enforcementAction)
 		totalViolationsPerEnforcementAction[ea]++
-		logViolation(am.log, r.Constraint, r.EnforcementAction, result)
+		logViolation(am.log, r.Constraint, r.EnforcementAction, rkind, rnamespace, rname, message)
 	}
 	// log constraints with violations
 	for link := range updateLists {
 		ar := updateLists[link][0]
 		logConstraint(am.log, ar.constraint, ar.enforcementAction, totalViolationsPerConstraint[link])
```

Comment:
Since `addAuditResponsesToUpdateLists` is called on every violating object now, we don't want to logConstraint on every call right?

Reply:
updated